sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -477,3 +498,204 @@ aes_error_t aes_gcm_gctr(const aes_key_len_t key_len, uint32_t *key_shares[2],
+ * @param j0 Destination for the output counter block
+ * @return OK or error
+ */
+static aes_error_t aes_gcm_counter(const size_t iv_len, const uint8_t *iv,
+                                   aes_gcm_product_table_t *tbl,
+                                   aes_block_t *j0) {
+  if (iv_len == 12) {
+    // If the IV is 96 bits, then J0 = (IV || {0}^31 || 1).
+    memcpy(j0->data, iv, iv_len);
+    // Set the last word to 1 (as a big-endian integer).
+    j0->data[kAesBlockNumWords - 1] = reverse_bytes(1);
+  } else if (iv_len == 16) {
+    // If the IV is 128 bits, then J0 = GHASH(H, IV || {0}^120 || 0x80).
+    memset(j0->data, 0, kAesBlockNumBytes);
+    aes_gcm_ghash_update(tbl, iv_len, iv, j0);
+    uint8_t buffer[kAesBlockNumBytes];
+    memset(buffer, 0, kAesBlockNumBytes);
+    buffer[kAesBlockNumBytes - 1] = 0x80;
+    aes_gcm_ghash_update(tbl, kAesBlockNumBytes, buffer, j0);
+  } else {
+    // Should not happen; invalid IV length.
+    return kAesInternalError;
+  }
+
+  return kAesOk;
+}
+
+/**
+ * Compute the AES-GCM authentication tag.
+ *
+ * AES parameters must be set to execute AES-CTR encryption, and have the key
+ * already loaded.
+ *
+ * @param aes_ctr_params Parameters for AES-CTR
+ * @param tbl Product table for the hash subkey H
+ * @param ciphertext_len Length of the ciphertext in bytes
+ * @param ciphertext Ciphertext value
+ * @param aad_len Length of the associated data in bytes
+ * @param aad Associated data value
+ * @param j0 Counter block (J0 in the NIST specification)
+ * @param tag Buffer for output tag (128 bits)
+ */
+aes_error_t aes_gcm_compute_tag(const aes_key_len_t key_len,
+                                uint32_t *key_shares[2],
+                                const aes_gcm_product_table_t *tbl,
+                                const size_t ciphertext_len,
+                                const uint8_t *ciphertext, const size_t aad_len,
+                                const uint8_t *aad, const aes_block_t *j0,
+                                uint8_t *tag) {
+  // Compute S = GHASH(H, expand(A) || expand(C) || len64(A) || len64(C))
+  // where:
+  //   * A is the aad, C is the ciphertext
+  //   * expand(x) pads x to a multiple of 128 bits by adding zeroes to the
+  //     right-hand side
+  //   * len64(x) is the length of x in bits expressed as a
+  //     big-endian 64-bit integer.
+
+  // Compute GHASH(H, expand(A) || expand(C)).
+  aes_block_t s;
+  memset(s.data, 0, kAesBlockNumBytes);
+  aes_gcm_ghash_update(tbl, aad_len, aad, &s);
+  aes_gcm_ghash_update(tbl, ciphertext_len, ciphertext, &s);
+
+  // Compute len64(A) and len64(C) by computing the length in *bits* (shift by
+  // 3) and then converting to big-endian.
+  uint32_t last_block[4];
+  last_block[0] = reverse_bytes(aad_len >> 29);
+  last_block[1] = reverse_bytes(aad_len << 3);
+  last_block[2] = reverse_bytes(ciphertext_len >> 29);
+  last_block[3] = reverse_bytes(ciphertext_len << 3);
```

Comment:
If we've already called `check_buffer_lengths()` at this point, we know that `aad_len <= UINT32_MAX`, so we could just say `last_block[0] = 0;`. If the ciphertext is the same length as the plaintext, we could also do `last_block[2] = 0`.

Reply:
Almost but not quite. `aad_len` is expressed in bytes, but the length here is supposed to be in _bits_, hence the shifting by 3/29 rather than 0/32.

Reply:
Agree about using `__builtin_bswap`, that's done now!

About using `char *` I'm not sure. When I grep the codebase we really seem to prefer `uint8_t` for bytes:
```shell
opentitan$ grep -Ir "char *" sw/device | wc -l
448
opentitan$ grep -Ir "uint8_t *" sw/device | wc -l
758
```
And when I look more closely at those results, almost all the `char *` usages are for strings, not for numerical data. I'm hesitant to deviate from that convention, and changing to `char *` for `aes_gcm_ghash_update` would require a lot more changes to `char *`, to the extent that I'd probably want to switch the whole file over to avoid being inconsistent.

Instead, for now I just inserted some `memcpy()` to get around the issue and added a comment explaining that it's to avoid violating strict aliasing. And I'm going to ask around a bit about this convention and whether it's worth using `char *` as standard instead.